Workaround for issue #44254 (Steam cannot connect to friends network)

mat8913 opened the pull request:
###### Motivation for this change

When /etc/localtime is a valid symlink but doesn't point directly to a zoneinfo file, Steam has trouble connecting to friends network. I've reported this upstream, maybe they'll fix it. But for now, there's this workaround.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
steam: Set TZ to detected timezone on startup

Workaround for issue #44254 (Steam cannot connect to friends network)

Ekleog commented on pkgs/games/steam/chrootenv.nix:
```diff
@@ -178,6 +178,12 @@ in buildFHSUserEnv rec {
+    # https://github.com/NixOS/nixpkgs/issues/44254
+    test -z ''${TZ+x} &&
+    new_TZ="$(readlink -f /etc/localtime | grep -o '(?<=/zoneinfo/).*$' -P)" &&
+    export TZ="$new_TZ"
```
Is there a reason for using an intermediate variable instead of directly exporting `TZ` as the result of the `readlink | grep` expansion?

mat8913 replied:
Yes. This way TZ is left alone if the `grep` fails.

Ekleog replied:
Oh indeed. Maybe add in a comment to warn future code readers not to do what I was thinking of? :)

Ekleog commented on pkgs/games/steam/chrootenv.nix:
```diff
@@ -178,6 +178,12 @@ in buildFHSUserEnv rec {
+    # Workaround for issue #44254 (Steam cannot connect to friends network)
+    # https://github.com/NixOS/nixpkgs/issues/44254
+    test -z ''${TZ+x} &&
+    new_TZ="$(readlink -f /etc/localtime | grep -o '(?<=/zoneinfo/).*$' -P)" &&
```
The `-P` argument to `grep` could go just after `grep` like `grep -P -o '...'`, that would likely make reading the code easier (as one would know that they are reading a PCRE regex before actually reading it)

mat8913 replied:
Thanks for the advice. I'll do that.

Commit:
steam workaround: Move -P argument next to grep

Makes reading the code easier, as one would know that they are reading a PCRE regex before actually reading it.

Commit:
steam workaround: Add comment explaining `new_TZ`

Ekleog reviewed (approved):
Great, thanks!

I think this is ready for merge :)

grahamc commented on pkgs/games/steam/chrootenv.nix:
```diff
@@ -178,6 +178,14 @@ in buildFHSUserEnv rec {
   profile = ''
+    # Workaround for issue #44254 (Steam cannot connect to friends network)
+    # https://github.com/NixOS/nixpkgs/issues/44254
+    test -z ''${TZ+x} &&
```
Please make this a proper if statement, the `&&` trickiness is a bit confusing.

Commit:
steam workaround: Convert && to if statements

Infinisil commented:
If somebody needs this to be backported, feel free to open another PR for it (with a cherry-pick -xe)
